python/cudf/cudf/core/frame.py

```diff
@@ -517,20 +543,25 @@ def _get_columns_by_index(self, indices):
         )
 
     def _gather(self, gather_map, keep_index=True, nullify=False):
+        """Gather rows of frame specified by indices in `gather_map`.
+
+        Set rows to null for all out of bound indices if nullify is `True`.
+        """
+        gather_map = cudf.core.column.as_column(gather_map)
         if not is_integer_dtype(gather_map.dtype):
             gather_map = gather_map.astype("int32")
-        result = self.__class__._from_data(
-            *libcudf.copying.gather(
-                self,
-                as_column(gather_map),
-                keep_index=keep_index,
-                nullify=nullify,
-            )
+
+        if not nullify and not _gather_map_is_valid(gather_map, len(self)):
+            raise IndexError("Gather map index is out of bounds.")
+
+        result = self.__class__._from_maybe_indexed_columns(
+            libcudf.copying.gather(
+                list(self._columns), gather_map, nullify=nullify,
+            ),
+            self._column_names,
         )
 
-        result._copy_type_metadata(self, include_index=keep_index)
-        if keep_index and self._index is not None:
-            result._index.names = self._index.names
+        result._copy_type_metadata(self)
```

Comment:
But .. `_from_maybe_indexed_columns` is a classmethod, how does it know anything about the type of source column?

Reply:
It's a classmethod because you just chose to make it one ;) but you are probably correct that there's a benefit to defining the classmethod for creating the object separately from the metadata handling. Perhaps it would be more appropriate to define _another_ (not class) method that wraps calls to `_from_columns` with additional instance-specific attributes. Something like:

```
def _from_columns_like_self(self, ...):
    obj = self._from_columns(...)
    # Do other things to postprocess obj
    return obj
```

We can wait to do this until later, so there's no need to clutter this PR. Consider this a suggestion for how this might look in the future. I would definitely encourage you to think about how we can define better patterns and restructure this code to maximize the chances that future devs copying existing snippets will have to copy the simplest possible snippet for things to "just work". Defining the right set of functions should make our code both self-documenting and more error-resistant (both user and developer error).